### ISSUE TITLE
"IllegalStateException: endPosTable already set" when adding apt output to sourcesets

### ISSUE THREAD

**Issue by fab1an** [tbroyer/gradle-apt-plugin#23]
I use [freebuilder](https://github.com/google/FreeBuilder) with the apt plugin.

I have 
```
    /* add apt to source dir */
    sourceSets.main.java.srcDir "$buildDir/generated/source/apt/main"
```

in my `build.gradle` so that my IDE will find the generated source files.

If I compile everything after a clean it works fine, but if I compile a second time, without doing a clean I get:

```
An exception has occurred in the compiler (1.8.0_77). Please file a bug against the Java compiler via the Java bug reporting page (http://bugreport.java.com) after checking the Bug Database (http://bugs.java.com) for duplicates. Include your program and the following diagnostic in your report. Thank you.
java.lang.IllegalStateException: endPosTable already set
        at com.sun.tools.javac.util.DiagnosticSource.setEndPosTable(DiagnosticSource.java:136)
        at com.sun.tools.javac.util.Log.setEndPosTable(Log.java:350)
        at com.sun.tools.javac.main.JavaCompiler.parse(JavaCompiler.java:667)
        at com.sun.tools.javac.main.JavaCompiler.parseFiles(JavaCompiler.java:950)
        at com.sun.tools.javac.processing.JavacProcessingEnvironment$Round.<init>(JavacProcessingEnvironment.java:892)
        at com.sun.tools.javac.processing.JavacProcessingEnvironment$Round.next(JavacProcessingEnvironment.java:921)
        at com.sun.tools.javac.processing.JavacProcessingEnvironment.doProcessing(JavacProcessingEnvironment.java:1187)
        at com.sun.tools.javac.main.JavaCompiler.processAnnotations(JavaCompiler.java:1170)
        at com.sun.tools.javac.main.JavaCompiler.compile(JavaCompiler.java:856)
...
```

When I don't add the source it works, but my IDE won't compile. Is there a way around this?

**Comment by tbroyer** [comment]
What's your IDE? Have you looked at the IDE section of the README?

**Comment by fab1an** [comment]
I use IntelliJ IDEA2016. I've looked at the Readme, the 2016 edition overrides the manually configured sourceset, when syncing from gradle.

**Comment by tbroyer** [comment]
What do you mean by "manually configured sourceset"? I haven't yet tried with 2016.1 but don't think it'll be different: workflow is a) import Gradle project (using IntelliJ support for Gradle, not `gradle idea`), b) enable annotation processing as described in the README; sync'ing again from Gradle should leave the annotation processing configuration alone.

BTW, the JavaC error you're seeing is due to the generated sources being part of the input files too (the way it's surfaced by JavaC is a bug in JavaC, but underlying cause is that outputs are inputs)

**Comment by fab1an** [comment]
I do it exactly like you say, I don't use the idea plugin at all.

I need to have `sourceSets.main.java.srcDir "$buildDir/generated/source/apt/main"` in my build.gradle, otherwise IDEA would overwrite the manually added source-set each time I sync.

**Comment by tbroyer** [comment]
Ah, you actually have to apply the `idea` plugin to get the appropriate configuration; you just won't run the associated tasks.
I really need to improve the docs.

**Comment by tbroyer** [comment]
Actually, I think the README was already clear enough about this, as the intro says: _“automatically configures IntelliJ IDEA and/or Eclipse **when the idea or eclipse plugins are applied**”_ (emphasis mine)

I added a note in the _Usage with IDEs_ section that the plugins are needed even when using the IDE integration: 1d0ac4e

**Comment by barisarican** [comment]
there is already bug reported regarding this issue 
https://bugs.openjdk.java.net/browse/JDK-8067747

I have also same issue, this thread might be helpful

https://github.com/immutables/immutables/issues/134

**Comment by Ramblurr** [comment]
So I have this issue when building from the CLI with gradle. If I don't clear the generated classes before doing a 2nd build, I trigger this bug.

Is there really no workaround?

**Comment by tbroyer** [comment]
@Ramblurr What issue do you have exactly? What's your configuration? (share your `build.gradle` in a gist)

The OP was a misconfiguration (feeding the output as input of the compiler – for the next round) due to a misunderstanding, trying to fix an IDE configuration the wrong way. If you're doing the same mistake, then this is your problem; then ask yourself why you did it in the first place, and fix that instead.
TL;DR: if you added something like `sourceSets.main.java.srcDir "$buildDir/generated/source/apt/main"` for your IDE to "see" the code generated by the annotation processors, then remove it and apply the IDE plugin instead (`apply plugin: "idea"` or `apply plugin: "eclipse"`, or the equivalent `plugins { id "idea" }` and `plugins { id "eclipse" }`). If that still doesn't work, then carefully review the section about IDEs in the README, and if that _still_ doesn't work, then file an issue (but please note that IDE support is provided as a best-effort only)

**Comment by eximius313** [comment]
I have exactly the same problem with Eclipse 4.6.3
Even if I apply and run ```apply plugin: "eclipse"``` eclipse still doesn't see classes generated by ```apt``` during test runtime. After adding ```sourceSets.main.java.srcDir "$buildDir/generated/source/apt/main"``` tests work fine, but instead this nasty ```java.lang.IllegalStateException: endPosTable already set``` occurs

**Comment by eximius313** [comment]
hmmm... I think I know what could cause this behaviour...
I'm using both - ```Metamodel Generator``` and ```Mapstruct```
```
apt (
	["org.hibernate:hibernate-jpamodelgen:5.2.9.Final"],
	["org.mapstruct:mapstruct-processor:1.2.0.Beta2"]
)
```
I can find both Metamodel and Mapstruct generated classes under Gradle's ```build\generated``` folder, but in ```.apt_generated``` there is only Metamodel. I think these is the reason of Eclipse not "seeing" Mapstruct, but why it is not generated?

**Comment by tbroyer** [comment]
@eximius313 Can you check the factory path in Project ➡ Properties ➡ Java Compiler ➡ Annotation Processing ➡ Factory Path ?

**Comment by eximius313** [comment]
![image](https://cloud.githubusercontent.com/assets/5578017/26313810/4f498e5e-3f0c-11e7-8569-103ab567f5a3.png)

**Comment by tbroyer** [comment]
@eximius313 Sorry for the delay. I copy/pasted your `apt` block into a dummy project:
```gradle
plugins {
  id 'java'
  id 'net.ltgt.apt' version '0.10'
  id 'idea'
  id 'eclipse'
}

repositories {
  mavenCentral()
}

dependencies {
  apt (
    ["org.hibernate:hibernate-jpamodelgen:5.2.9.Final"],
    ["org.mapstruct:mapstruct-processor:1.2.0.Beta2"]
  )
}
```
and invoked `./gradlew eclipseFactorypath` and the generated `.factorypath` correctly references the 2 processors, along with the `jboss-logging` dependency:
```xml
<factorypath>
  <factorypathentry kind='EXTJAR' id='/mnt/ssd/tbr/.gradle/caches/modules-2/files-2.1/org.hibernate/hibernate-jpamodelgen/5.2.9.Final/7860a0e884972c9919a29803585f89fb46016c06/hibernate-jpamodelgen-5.2.9.Final.jar' enabled='true' runInBatchMode='false' />
  <factorypathentry kind='EXTJAR' id='/mnt/ssd/tbr/.gradle/caches/modules-2/files-2.1/org.mapstruct/mapstruct-processor/1.2.0.Beta2/5eb4495e9e78d7eef8c5a5a90813ee2edfed1fdc/mapstruct-processor-1.2.0.Beta2.jar' enabled='true' runInBatchMode='false' />
  <factorypathentry kind='EXTJAR' id='/mnt/ssd/tbr/.gradle/caches/modules-2/files-2.1/org.jboss.logging/jboss-logging/3.3.0.Final/3616bb87707910296e2c195dc016287080bba5af/jboss-logging-3.3.0.Final.jar' enabled='true' runInBatchMode='false' />
</factorypath>
```
Here's the output of `./gradlew dependencies --configuration=apt`:
```
:dependencies

------------------------------------------------------------
Root project
------------------------------------------------------------

apt - Processor path for source set 'main'
+--- org.hibernate:hibernate-jpamodelgen:5.2.9.Final
|    \--- org.jboss.logging:jboss-logging:3.3.0.Final
\--- org.mapstruct:mapstruct-processor:1.2.0.Beta2

BUILD SUCCESSFUL

Total time: 0.557 secs
```
This is with Gradle 3.5.

Your factory path references Lombok and `org.eclipse.jst.ws.annotations.core`, which means it's not the one generated by `./gradlew eclipseFactorypath`.
